dm/syncer/syncer_test.go

```diff
@@ -1703,3 +1687,72 @@ func (s *testSyncerSuite) TestExecuteSQLSWithIgnore(c *C) {
 
 	c.Assert(mock.ExpectationsWereMet(), IsNil)
 }
+
+func genDefaultSubTaskConfig4Test() *config.SubTaskConfig {
+	loaderDir, err := os.MkdirTemp("", "loader")
+	if err != nil {
+		panic(err) // no happen
+	}
+
+	loaderCfg := config.LoaderConfig{
+		Dir: loaderDir,
+	}
+	cfg := &config.SubTaskConfig{
+		From:             config.GetDBConfigForTest(),
+		To:               config.GetDBConfigForTest(),
+		ServerID:         101,
+		MetaSchema:       "test",
+		Name:             "syncer_ut",
+		ShadowTableRules: []string{config.DefaultShadowTableRules},
+		TrashTableRules:  []string{config.DefaultTrashTableRules},
+		Mode:             config.ModeIncrement,
+		Flavor:           "mysql",
+		LoaderConfig:     loaderCfg,
+		UseRelay:         false,
+	}
+	cfg.Experimental.AsyncCheckpointFlush = true
+	cfg.From.Adjust()
+	cfg.To.Adjust()
+	return cfg
+}
+
+func TestWaitBeforeRunExit(t *testing.T) {
```

Comment:
this is not an unit test covers https://github.com/pingcap/tiflow/pull/4035#discussion_r793311221

What I want is only to test `waitBeforeRunExit`, not whole Syncer. I think you can construct a simple Syncer, call `waitBeforeRunExit(ctx1)`, cancel ctx1, wait `maxPauseOrStopWaitTime`, check if `runCtx` is canceled

Reply:
add this test  in https://github.com/pingcap/tiflow/pull/4035/commits/1e51bec3be4a8460970648afd7c163ee516a77a4

and i thinks we can keep old test(// test s.Run will not exit unit caller cancel ctx or call s.runCancel)